library/aes.c

```diff
@@ -851,6 +853,11 @@ int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
         return( mbedtls_aesni_crypt_ecb( ctx, mode, input, output ) );
 #endif
 
+#if defined(MBEDTLS_ARMV8CE_AES_C)
+	// We don't do runtime checking for ARMv8 Crypto Extensions
```

Comment:
You almost certainly don't want to go to /proc/cpuinfo... To detect processor features in runtime, use [hwcaps](https://www.kernel.org/doc/html/latest/arm64/elf_hwcaps.html) instead:

```
#include <asm/hwcap.h>
#include <sys/auxv.h>

int crypto_extensions_present(void)
{
        unsigned long hwcaps = getauxval(AT_HWCAP);
        return (hwcaps & HWCAP_AES) && (hwcaps & HWCAP_PMULL);
}
```

Reply:
@pawelmoll Thanks! How portable is this in terms of kernel, libc, toolchain, versions? Can we assume it's always there when compiling for Aarch64/Aarch32 or should we provide a compile-time option?

Reply:
According to its man page:
> The getauxval() function was added to glibc in version 2.16.
(which, it seems, was released in 2012).

As to hwcaps themselves, AArch64 (aka arm64) kernels were providing them from the day one :-) (or at least early enough that you wouldn't care).

As to AArch32/arm environment... It seems that the AES and MULL got into AT_HWCAP2 word:
(https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/include/uapi/asm/hwcap.h)
but I must say that I have no crypto-enabled platform I could test it on...

Reply:
Re AArch32 again: they were added "only" in 5.3
(https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/arch/arm/include/uapi/asm/hwcap.h?id=8258a9895c99cdaacad8edc4748c0a624c710961)

Reply:
So that's Linux-only, not some portability layer? At this point, for Arch64, we can probably reasonably assume that `__linux__` implies ≥5.3. Should we also have a guard on `__GLIBC__`?

Reply:
It's definitely a Linux thing, yes. And yes, on AArch64 I'd just assume HW_CAPS have all what you need (checked and the AES and MULL bits were added in 3.13).

As to "non-glibc libc-s", musl and bionic seem to have the function:
http://git.musl-libc.org/cgit/musl/tree/src/misc/getauxval.c
https://android.googlesource.com/platform/bionic/+/master/libc/bionic/getauxval.cpp
but, for example, uclibc doesn't...

Reply:
Also, there's [an old lwn article describing auxv](https://lwn.net/Articles/519085/):
> Even without using the new library function, an application that wants to access the auxiliary vector merely needs to obtain the address of the location that follows the NULL pointer at the end of the environment list.